Thank you @geoand.

So the reason why this PR was failing on Windows was due to the following exception (which wasn't accessible easily in the CI infrastructure)

```
Caused by: java.nio.file.FileSystemException: C:\quarkus\integration-tests\maven\target\test-classes\projects\project-uberjar-true\target\acme-1.0-SNAPSHOT-runner.jar: The process cannot access the file because it is being used by another process.

	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
	at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
	at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
	at java.nio.file.Files.deleteIfExists(Files.java:1165)
	at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildUberJar(JarResultBuildStep.java:169)
	at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildRunnerJar(JarResultBuildStep.java:138)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:948)
	... 7 more
```
This was because the newly added method `ensureManifestOfJarIsReadableByJarInputStream` method wasn't closing the streams it was opening. I've updated this PR to fix that. This change passed locally on a Windows setup. So let's see how it goes now on our CI.

As for the changes in this PR itself, they look good to me.

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
@@ -182,6 +182,11 @@ public class JarResultBuildStep {
 
             final List<AppDependency> appDeps = curateOutcomeBuildItem.getEffectiveModel().getUserDependencies();
 
+            AppArtifact appArtifact = curateOutcomeBuildItem.getEffectiveModel().getAppArtifact();
+            // the manifest needs to be the first entry in the jar, otherwise JarInputStream does not work properly
+            // see https://bugs.openjdk.java.net/browse/JDK-8031748
+            generateManifest(runnerZipFs, classPath.toString(), packageConfig, appArtifact, applicationInfo);
+
             for (AppDependency appDep : appDeps) {
                 final AppArtifact depArtifact = appDep.getArtifact();
                 final Path resolvedDep = depResolver.resolve(depArtifact);
@@ -261,12 +266,8 @@ public class JarResultBuildStep {
                     }
                 }
             }
-
             copyCommonContent(runnerZipFs, services, applicationArchivesBuildItem, transformedClasses, generatedClasses,
                     generatedResources, seen);
-            AppArtifact appArtifact = curateOutcomeBuildItem.getEffectiveModel().getAppArtifact();
-            generateManifest(runnerZipFs, classPath.toString(), packageConfig, appArtifact, applicationInfo);
-
         }
 
         runnerJar.toFile().setReadable(true, false);
@@ -375,10 +376,13 @@ public class JarResultBuildStep {
         final List<AppDependency> appDeps = curateOutcomeBuildItem.getEffectiveModel().getUserDependencies();
 
         copyLibraryJars(transformedClasses, libDir, depResolver, classPath, appDeps);
+
+        AppArtifact appArtifact = curateOutcomeBuildItem.getEffectiveModel().getAppArtifact();
+        // the manifest needs to be the first entry in the jar, otherwise JarInputStream does not work properly
+        // see https://bugs.openjdk.java.net/browse/JDK-8031748
+        generateManifest(runnerZipFs, classPath.toString(), packageConfig, appArtifact, applicationInfo);
         copyCommonContent(runnerZipFs, services, applicationArchivesBuildItem, transformedClasses, allClasses,
                 generatedResources, seen);
-        AppArtifact appArtifact = curateOutcomeBuildItem.getEffectiveModel().getAppArtifact();
-        generateManifest(runnerZipFs, classPath.toString(), packageConfig, appArtifact, applicationInfo);
     }
 
     private void copyLibraryJars(TransformedClassesBuildItem transformedClasses, Path libDir, AppModelResolver depResolver,
@@ -528,6 +532,8 @@ public class JarResultBuildStep {
                 manifest.read(is);
             }
             Files.delete(manifestPath);
+        } else {
+            Files.createDirectories(runnerZipFs.getPath("META-INF"));
         }
         Attributes attributes = manifest.getMainAttributes();
         attributes.put(Attributes.Name.MANIFEST_VERSION, "1.0");

--- a/integration-tests/maven/src/test/java/io/quarkus/maven/it/PackageIT.java
+++ b/integration-tests/maven/src/test/java/io/quarkus/maven/it/PackageIT.java
@@ -4,19 +4,24 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import java.io.File;
+import java.io.FileInputStream;
 import java.io.FileNotFoundException;
+import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Properties;
 import java.util.Set;
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
 
 import org.apache.maven.shared.invoker.MavenInvocationException;
-import org.assertj.core.util.Arrays;
 import org.junit.Assert;
 import org.junit.jupiter.api.Test;
 
@@ -30,7 +35,7 @@ public class PackageIT extends MojoTestBase {
 
     @Test
     public void testPackageWorksWhenUberjarIsFalse()
-            throws MavenInvocationException, FileNotFoundException, InterruptedException {
+            throws MavenInvocationException, IOException, InterruptedException {
         testDir = initProject("projects/uberjar-check", "projects/project-uberjar-false");
 
         running = new RunningInvoker(testDir, false);
@@ -40,12 +45,24 @@ public class PackageIT extends MojoTestBase {
         assertThat(result.getProcess().waitFor()).isEqualTo(0);
 
         final File targetDir = getTargetDir();
-        assertThat(getNumberOfFilesEndingWith(targetDir, ".jar")).isEqualTo(2);
+        List<File> jars = getFilesEndingWith(targetDir, ".jar");
+        assertThat(jars).hasSize(2);
+
+        // make sure the jar can be read by JarInputStream
+        ensureManifestOfJarIsReadableByJarInputStream(
+                jars.stream().filter(f -> f.getName().contains("-runner")).findFirst().get());
+    }
+
+    private void ensureManifestOfJarIsReadableByJarInputStream(File jar) throws IOException {
+        try (InputStream fileInputStream = new FileInputStream(jar)) {
+            Manifest manifest = new JarInputStream(fileInputStream).getManifest();
+            assertThat(manifest).isNotNull();
+        }
     }
 
     @Test
     public void testPackageWorksWhenUberjarIsTrue()
-            throws MavenInvocationException, FileNotFoundException, InterruptedException {
+            throws MavenInvocationException, IOException, InterruptedException {
         testDir = initProject("projects/uberjar-check", "projects/project-uberjar-true");
 
         createAndVerifyUberJar();
@@ -53,7 +70,7 @@ public class PackageIT extends MojoTestBase {
         createAndVerifyUberJar();
     }
 
-    private void createAndVerifyUberJar() throws FileNotFoundException, MavenInvocationException, InterruptedException {
+    private void createAndVerifyUberJar() throws IOException, MavenInvocationException, InterruptedException {
         Properties p = new Properties();
         p.setProperty("quarkus.package.uber-jar", "true");
 
@@ -63,8 +80,11 @@ public class PackageIT extends MojoTestBase {
         assertThat(result.getProcess().waitFor()).isEqualTo(0);
 
         final File targetDir = getTargetDir();
-        assertThat(getNumberOfFilesEndingWith(targetDir, ".jar")).isEqualTo(1);
+        List<File> jars = getFilesEndingWith(targetDir, ".jar");
+        assertThat(jars).hasSize(1);
         assertThat(getNumberOfFilesEndingWith(targetDir, ".original")).isEqualTo(1);
+
+        ensureManifestOfJarIsReadableByJarInputStream(jars.get(0));
     }
 
     @Test
@@ -105,7 +125,7 @@ public class PackageIT extends MojoTestBase {
      */
     @Test
     public void testRunnerUberJarHasValidCRC() throws Exception {
-        testDir = initProject("projects/uberjar-check", "projects/project-uberjar-true");
+        testDir = initProject("projects/uberjar-check", "projects/project-uberjar-true2");
 
         running = new RunningInvoker(testDir, false);
 
@@ -134,7 +154,7 @@ public class PackageIT extends MojoTestBase {
      */
     @Test
     public void testRunnerJarHasValidCRC() throws Exception {
-        testDir = initProject("projects/uberjar-check", "projects/project-uberjar-false");
+        testDir = initProject("projects/uberjar-check", "projects/project-uberjar-false2");
 
         running = new RunningInvoker(testDir, false);
         final MavenProcessInvocationResult result = running.execute(Collections.singletonList("package"),
@@ -150,9 +170,13 @@ public class PackageIT extends MojoTestBase {
         assertZipEntriesCanBeOpenedAndClosed(runnerJar);
     }
 
-    private int getNumberOfFilesEndingWith(File dir, String suffix) {
+    private List<File> getFilesEndingWith(File dir, String suffix) {
         final File[] files = dir.listFiles((d, name) -> name.endsWith(suffix));
-        return files != null ? files.length : 0;
+        return files != null ? Arrays.asList(files) : Collections.emptyList();
+    }
+
+    private int getNumberOfFilesEndingWith(File dir, String suffix) {
+        return getFilesEndingWith(dir, suffix).size();
     }
 
     private File getTargetDir() {

